### ISSUE TITLE
systemd tarballs should be signed

### ISSUE THREAD

**Issue by keszybz** [systemd/systemd#2926]
For some background discussion, see e.g. https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/2TBK4LLNRH73QJQSXWFPCQYHGTSJ3C7P/.

We currently use autogenerated github tarballs. I think this is convenient, and we can keep that. But we should provide a signature for this tarball, so that downstream maintainers can easily check that the downloaded code is legitimate. The signatures would be generated for tagged versions only.

Proposal:
- when making a release, also generate the tarball locally and create a detached signature for the github tarball
- upload the signature file to github

Something like this:

```
v=229; git archive --prefix=systemd-$v/ v$v | gzip --no-name > systemd-$v.tar.gz
gpg2 --detach-sign systemd-$v.tar.gz 
curl -X POST -H 'Content-Type: text/plain' https://github.com/repos/:systemd/:systemd/releases/:v$v/assets?name=systemd-$v.tar.gz.gpg --data-binary systemd-$v.tar.gz.gpg
```

(Last part untested)

Or we could generate proper .xz compressed tarballs and sign that, since we're already mucking around with uploads.

**Comment by poettering** [comment]
This sounds like a step back, no?

The idea is that we sign git tags, and do not provide tarballs at all anymore. If distros want tarballs, they can use the compat stuff that allows pulling down git versions as tarballs, but that should be compat really, and the suggested form of retrieval should be pulling directly from git.

Or to turn this around: this was really supposed to be an excercise in cutting cruft and simplifying the release generation for us by dumbing it down to a single "git tag", and nothing else.

gpg signed tarballs is so ten years ago, let's not polish this legacy...

**Comment by keszybz** [comment]
Yes, you can always verify the code by cloning the repo and checking the tag. But in practice people tend to download the tarball, because it's easier and more convenient if you're doing a one-off build. In particular Fedora rpm infrastructure is based on tarballs. It's something that I'd love to change, but that would be a multi-year project. I think that for now, for purely practical reasons, we should sign the tarballs, to make it easy for people do be secure when using the most common distribution channel.

**Comment by keszybz** [comment]
@poettering would you mind if I just upload signatures for the tarballs after the fact?

**Comment by poettering** [comment]
Sure. Go ahead. But are you sure that the git archive output is stable? I.e. that you always get the excact bytewise identical tarball?

**Comment by keszybz** [comment]
Yes, it's stable. For example this gets regularly testing in Fedora, when people specify the github tarball as Source0, and add the hash of the tarball to `sources` file. If it wasn't stable, there would be a lot of warnings.

**Comment by shibumi** [comment]
Hello,
Anything new about this issue? @keszybz  are you signing the systemd tarballs in future? :) Just asking, because it would be awesome if somebody does that.

**Comment by keszybz** [comment]
I made a little investigation today: both tar.gz and .zip output is stable and predictable (mostly). It's trivial to generate .tar.gz output that is identical to the github provided tarball:
```
i=233; git archive -o v$i.tar.gz --prefix=systemd-$i/ v$i
```
But with .zip it's more complicated, because some strange timezone is embedded in the github-generated files. This
```
for i in {224..233}; do TZ=UTC+7 git archive -o v$i.zip --prefix=systemd-$i/ v$i;done
```
gives me output which is identical for most releases, but sometimes there's a mismatch:
```
systemd-224.tar.gz: OK
systemd-225.tar.gz: OK
systemd-226.tar.gz: OK
systemd-227.tar.gz: OK
systemd-228.tar.gz: OK
systemd-229.tar.gz: OK
systemd-230.tar.gz: OK
systemd-231.tar.gz: OK
systemd-232.tar.gz: OK
systemd-233.tar.gz: OK
systemd-224.zip: OK
systemd-225.zip: OK
systemd-226.zip: OK
systemd-227.zip: OK
systemd-228.zip: FAILED
systemd-229.zip: FAILED
systemd-230.zip: OK
systemd-231.zip: OK
systemd-232.zip: OK
```
Diffoscope says:
```
...
│ @@ -74392,15 +74392,15 @@
│    version of encoding software:                   2.3
│    minimum file system compatibility required:     MS-DOS, OS/2 or NT FAT
│    minimum software version required to extract:   1.0
│    compression method:                             deflated
│    compression sub-type (deflation):               normal
│    file security status:                           not encrypted
│    extended local header:                          no
│ -  file last modified on (DOS date/time):          2015 Nov 18 00:59:06
│ +  file last modified on (DOS date/time):          2015 Nov 17 23:59:06
│    file last modified on (UT extra field modtime): 2015 Nov 18 07:59:06 local
│    file last modified on (UT extra field modtime): 2015 Nov 18 07:59:06 UTC
│    32-bit CRC value (hex):                         fd8cff10
│    compressed size:                                132 bytes
│    uncompressed size:                              198 bytes
│    length of filename:                             35 characters
│    length of extra field:                          9 bytes
```
So there's a mismatch when the release is around midnight. Might be a different timezone, or some bug in their generation of zips, or something else. I suppose people are unlikely to use .zip files, so it'd be enough to sign the .tar.gz file.

So for now, I'll just upload signatures for the auto-generated .tar.gz files.

In the future we might want to consider uploading a .tar.xz file that we generate locally, along with a signature. One advantage would be smaller size: .xz gives is about 50% smaller than the .zip, and about 30% smaller than the .tar.gz. I have a script to upload stuff to github that I made for another project [1], but it doesn't work fully out-of-the-box, because the github release must be created by hand (i.e. making of the tag is not enough, it is also necessary to provide some description).

[1] https://github.com/neurord/stochdiff/blob/master/util/upload-tarball.sh

**Comment by keszybz** [comment]
OK, I now uploaded signatures for a bunch of releases to github (*), and it *does* feel like a step backward. It'd be much nicer to upload the signature along with a proper tarball with a proper name and proper compression, all at once.

(*) This just means that the tarballs I generated locally from my own copy of the repo match what you can download from github. Whatever that is worth.

**Comment by NicoHood** [comment]
I've discussed the topic in https://github.com/systemd/systemd/issues/6295 . You can easily achieve that by using [gpgit](https://github.com/NicoHood/gpgit) for example. It generated a local tarball, signs it and uploads it to Github. It would increase the security of the tarballs a lot, if they are generated from the local git archive. Thanks.

**Comment by keszybz** [comment]
For now I'll just keep signing the automatically generated tarballs (or rather a local copy that is identical the the github-generated one). We only do a release every 3 months or so, so the extra manual step doesn't really matter.

gpgit looks very nice. It would be great if it was packaged. There seems to be an arch package, but nothing for Fedora or Debian, and it's hard to depend on it without that.

**Comment by NicoHood** [comment]
@keszybz thanks for the feedback. You can install gpgit dependencies via pip and the tool itself via simple download, its a single file. In the future I want to publish it also via pip, so the installation is simple on all distributions. I have to look into this.

About systemd, you understood my point, so its up to you now how you continue. Thanks.

**Comment by eli-schwartz** [comment]
In Arch Linux, we don't really care about whether the sources are downloaded over git:// or https:// -- our packaging tools support either one.

Something we do care about, however, is whether the systemd-stable commits, which we often use instead of tags in order to more conveniently obtain important backports, are signed. Currently, these commits are *not* signed, and therefore in order to use systemd-stable, we must give up on PGP verification.

This also has wider applicability to anyone wishing to build their own systemd from git, or perhaps to use the [Arch Linux AUR package](https://aur.archlinux.org/packages/systemd-git) that automates that process.

Please consider setting `git config --global commit.gpgsign true` on your local machines, and enforcing a general policy where this must be done in order to push to master. Note that for proper effect you'll need to also merge PRs via the command line rather than Github's interface. (`git merge --no-ff` followed by pushing to master will cause GitHub to close the commit with the expected "merged commit XXXXXX into master" status, and all PRs can be conveniently checked out by pulling `refs/pull/$number/head` from GitHub.)

**Comment by eli-schwartz** [comment]
> I think one of the consensus being made is that we should switch to tarballs automatically generated or even VCS package sources where appropriate, rather than relying on tarballs signed by developers and uploaded manually.

- incorrect
  - there are efforts to make it easy to validate that autoconf dist tarballs match what can be generated in a clean environment
  - but particularly since as a meson-based project, systemd can rest comfortably knowing that tarballs made with `meson dist` and uploaded by developers have the same contents as the automatically generated tarballs modulo tar metadata, and diffing them will reveal ~no differences. It is definitely easy to vet
- even if it were correct, there is no rule that says you cannot PGP-sign the automatically generated tarballs, and upload the PGP signature to github releases with the expectation that it has to match the tarball that github produces.


I would also like to note regarding VCS package sources, that not all distros can use this at all. For distros where tarballs are "mandatory, no discussion no debates end of story", the lack of tarball signatures is unfortunate and means no cryptographic verification is being done.

Sure, this is the status quo for like a decade now, but that doesn't mean the people who are missing out on that have any opinion other than "we would still like it if one day systemd would finally start doing it".

**Comment by eli-schwartz** [comment]
8 years and as far as I can tell no argument has been made *against* uploading PGP signatures to the github release other than an unqualified "so ten years ago" which isn't an argument -- it is an https://en.wikipedia.org/wiki/Appeal_to_ridicule

**Comment by YHNdnzj** [comment]
> 8 years and as far as I can tell no argument has been made *against* uploading PGP signatures to the github release other than an unqualified "so ten years ago" which isn't an argument -- it is an https://en.wikipedia.org/wiki/Appeal_to_ridicule

Honestly, there's no need to overreact like that. I'm _not_ against it either, but just wondering if there's still enough interest in this.

**Comment by eli-schwartz** [comment]
I'm not blaming you -- but this bug has been quiescent for the better part of a decade after someone other than you did exactly what I described. :)
